doc/architectural/CBMC-developer-guide.md

```diff
@@ -90,101 +275,242 @@ The `–help` option gives instructions for use and the
+FIXME
 
-  4 -> 5;
+## Exprt ##
+Exprt is the class to represent an expression. It inherits from irept,
+and the only things it adds to it are that every exprt has a named sub
+containing its type and everything in the sub of an exprt is again an
+exprt, not just an irept. You can think of exprt as a node in the
+abstract syntax tree for an expression. There are many classes that
+inherit from exprt and which represent more specific things. For
+example, there is minus_exprt, which has a sub of size 2 (for the two
+argument of minus).
 
-  1 [label="command line\nparsing" URL="\ref cbmc_parse_optionst"];
-  2 [label="preprocessing,\nparsing" URL="\ref preprocessing"];
-  3 [label="language\ntype-checking" URL="\ref type-checking"];
-  4 [label="goto\nconversion" URL="\ref goto-conversion"];
-  5 [label="instrumentation" URL="\ref instrumentation"];
-  6 [label="symbolic\nexecution" URL="\ref symbolic-execution"];
-  7 [label="SAT/SMT\nencoding" URL="\ref sat-smt-encoding"];
-  8 [label="decision\nprocedure" URL="\ref decision-procedure"];
-  9 [label="counter example\nproduction" URL="\ref counter-example-production"];
-}
-\enddot
+Recall that every irept has one piece of data of its own, i.e. its
+id(), and all other information is in its named_sub, comments_sub or
+sub. For exprts, the id() is used to specify why kind of exprt this is,
+so a minus_exprt has ID_minus as its id(). This means that a
+minus_exprt can be passed wherever an exprt is expected, and if you
+want to check if the expression you are looking at is a minus
+expression then you have to check its id().
 
-The \ref cbmc-user-manual "CBMC User Manual" describes CBMC from a user
-perspective. Each node in the diagram above links to the appropriate
-class or module documentation, describing that particular stage in the
-CBMC pipeline.
-CPROVER is structured in a similar fashion to a compiler. It has
-language specific front-ends which perform limited syntactic analysis
-and then convert to an intermediate format. The intermediate format can
-be output to files (this is what `goto-cc` does) and are (informally)
-referred to as “goto binaries” or “goto programs”. The back-end are
-tools process this format, either directly from the front-end or from
-it’s saved output. These include a wide range of analysis and
-transformation tools (see \ref other-tools).
+## Codet
+Exprts represent expressions and codets represent statements. Codet
+inherits from exprt, so all codets are exprts, with id() ID_code.
+Many different kinds of statements inherit from codet, and they are
+distinguished by their statement(). For example, a while loop would be
+represented by a code_whilet, which has statement() ID_while.
+Code_whilet has two operands in its sub, and helper functions to make
+it easier to use: cond() returns the first entry in its sub(), which
+is the condition for the while loop, and body() returns the second
+entry in its sub, which is the body of the while loop - this has to be
```

Comment:
Worth defining and then using the terminology "first subexpression, second subexpression, ..." rather than "... in its sub"?

Reply:
Use backticks for code elements